### PR TITLE
add dummy-ema-optimizer for model averaing

### LINKED ISSUE #3982: checkpoint merge & model averaging via dummy optimizer
I mentioned about EMA checkpoint (or model averaging) times ago here https://github.com/pytorch/torchtitan/issues/2355  and i dont see any related updates or new pr try to support it in torchtitan.

and i saw there is multiple optimizers support PR is merged https://github.com/pytorch/torchtitan/pull/3269 and i wonder that why dont we just add a dummy optimizer and make it state's as EMA weights? so all parallel stuff can essentially resolved automatically, and the core code is also gonna be simple, we can even borrow adamw's fused update there, and even cpu offload stuff.  and async ckpt saving can also handle the checkpoint saving ...   even though this solution maybe tricky to do smth like run val loss on EMA weightsetc but the solution itself is very easy to go.

i wonder like to discusses in brief see if this is a good idea to do then i can open PR (i think this should be easy to do). 

----

about why we are interested in model averaging.  so most of now day's llm trianing go with stable-decay LR scheduler, and one problem is, you cannot see model's actually performance in stable LR regime on-real-time.   (so you see model performance hugely improved in decay phase) and this is annoying.    and model averaging is the solution for that.   In stable-LR phase, instead of checking eval on current weights,  the averaged EMA weights can better measure the model's performance. 

And the extension of it will be scheduler-free, so one can essentially train LLM with constant LR and just check the EMA weights to get its actual performance.

### PR DESCRIPTION
**Pull request by rakkit**
add EMA implement for https://github.com/pytorch/torchtitan/issues/3982  

## What it is

`EMA` subclasses `OptimizersContainer` and behaves like a gradient-free pseudo-optimizer:
it holds one `ema_params` tensor per tracked tensor and updates them in place after each
real optimizer step. 

Two decay formulas:

1. **fixed** — `ema = decay * ema + (1 - decay) * param`
2. **dynamic** (default) — `decay = 2 ** (-1 / (half_life_fraction * num_updates))`, i.e.
   roughly 5% tail averaging ([reference](https://arxiv.org/abs/2602.03702v1))

## How to use

in Config
```python
ema=EMA.Config(half_life_fraction=0.05)
```

or from CLI
```bash
ema:config --ema.half-life-fraction 0.05 --ema.update-every-n-steps 1
```

convert EMA weights to HF checkpoint
```bash
scripts/checkpoint_conversion/convert_ema_to_hf.py
```
 which takes the same arguments as `convert_to_hf.py`. 

**Buffers.** MoE's `expert_bias_E` is a `register_buffer` updated by a non-gradient
load-balancing heuristic, so it is not a parameter and would not be tracked by EMA. To expliclty include them, we could set`buffer_patterns=["expert_bias_E$"]` folds it into the same `"ema"`
key.

**CPU offload.** `offload_to_cpu` runs the update on a side stream with an H2D/D2H
pipeline. Scratch is carved from a single flat buffer and reused chunk by chunk, so GPU
scratch stays bounded instead of growing to 1x parameter memory.

## Edge cases and semantics (some edge cases from claude, summarized by claude)

These are the cases worth reviewing; most of the non-obvious code is here.

**The decay schedule has to survive a resume.** `num_updates` is stateless, so derived from the trainer
step, not kept as a counter. Only `ema_params` are checkpointed, so a stored counter would
restart at 0 after a resume, making `decay ~ 2**-20` and overwriting the restored average
with the current weights on the very first firing.

**Turning EMA on mid-run.** An existing checkpoint has no `"ema"` key, so the load fails
until you exclude it once:

```python
checkpointer=CheckpointManager.Config(exclude_from_loading=["ema"])
```
The EMA then cold-starts from the just-loaded weights. 

**Model-only loads.** An HF safetensors load or an `initial_load_path` warm start never
restores EMA state, so it reseeds from the loaded weights the same way. Conversely,
resuming with EMA disabled against a checkpoint that has EMA data is fine: the key is
simply never requested.

**The final checkpoint has no EMA state.** `last_save_model_only` defaults to `True`, so
the last-step checkpoint is model weights only. Export the EMA from an interval
checkpoint, or set `last_save_model_only=False`. -> maybe we should fix this?

**Fused parameters.** EMA state is keyed by `named_parameters()` FQNs, which for fused
modules differ from the model state-dict keys the save hooks split them into
(`qkv_linear.wqkv` -> `wq`/`wk`/`wv`, `feed_forward.w13` -> `w1`/`w3`). One checkpoint
therefore holds `ema.state.<fused fqn>.ema_params` next to model keys `...wq.weight`. That
is the same contract optimizer state already follows. The conversion script overlays
through the module rather than the dict for this reason — writing EMA values into a
state-dict-keyed dict would add unmatched keys that `to_hf` silently drops, exporting the
trained weights instead.

**Tensors EMA does not track.** Only parameters with `requires_grad` are tracked, plus the
buffers matched by `buffer_patterns`. Everything else — frozen parameters, unmatched
buffers — is exported from the trained weights, so the HF result is a complete, loadable
checkpoint rather than one with missing keys. The conversion script tracks exactly the
FQNs the checkpoint holds, so a run that froze some parameters converts correctly.

**The tracked set is fixed at construction.** State is keyed by tensor identity, so
unfreezing a parameter, or replacing a parameter or buffer, after the EMA is built raises a
clear error instead of a bare `KeyError`. Freezing a parameter afterwards is fine; it just
stops being averaged.

**DTensor and pinned memory.** DTensor has no `pin_memory()` (NYI:
`aten._pin_memory.default`), so the offload path pins the local shard and rewraps it around
the parameter's live sharding at save/load time. There is a TODO for this, including the
process-group-restore concern.

### CODE REVIEW

File: torchtitan/components/checkpoint.py [hunk 1]
```diff
@@ -1070,6 +1082,36 @@ def _states_to_load(self, model_only: bool) -> dict[str, Any]:
...
+        ``prefix``. Used to detect an optional component that is genuinely
+        absent from a checkpoint before attempting to load it.
+        """
+        try:
+            metadata = dcp.FileSystemReader(checkpoint_id).read_metadata()
+            return any(k.startswith(prefix) for k in metadata.state_dict_metadata)
+        except Exception:
+            return False
+
+    def _maybe_exclude_missing_ema(
```
Comment (rakkit) [inline torchtitan/components/checkpoint.py:1096]: its mainly for BP and deal with edge case.
people may resume or doing continue training where EMA weights is missing there

File: torchtitan/components/ema.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        self.step_bias = config.step_bias
+        self.update_every_n_steps = config.update_every_n_steps
+        self.offload_to_cpu = config.offload_to_cpu
+        self.model_parts = model_parts
+
+        self.optimizers: list[_EMAParamOptimizer] = []
+        all_params: list[nn.Parameter] = []
+        for model in model_parts:
+            named_params = [
+                (name, p) for name, p in model.named_parameters() if p.requires_grad
```
Comment (rakkit) [inline torchtitan/components/ema.py:109]: for now, only learnable params will be included in EMA

File: torchtitan/components/ema.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        """Used when decay is None: decay = 2 ** (-1 / (half_life_fraction * t)),
+        where t is elapsed steps since start_step. Keeps roughly the most
+        recent half_life_fraction share of steps dominant. Default 0.05
+        matches the common decay = 2 ** (-20 / t) rule of thumb."""
+
+        start_step: int = 0
+        """First training step (matching Trainer.step) at which EMA tracking
+        begins. Intentionally decoupled from the LR scheduler's WSD phases."""
+
+        step_bias: int = 0
```
Comment (rakkit) [inline torchtitan/components/ema.py:82]: we need this because for mid-train, continues trainig or any cases people reset training step to 1, but LR scheduler or EMA scheduler is kept.

File: torchtitan/components/ema.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        per firing. If None (default), decay is instead computed dynamically
+        from half_life_fraction -- see below."""
+
+        half_life_fraction: float = 0.05
+        """Used when decay is None: decay = 2 ** (-1 / (half_life_fraction * t)),
+        where t is elapsed steps since start_step. Keeps roughly the most
+        recent half_life_fraction share of steps dominant. Default 0.05
+        matches the common decay = 2 ** (-20 / t) rule of thumb."""
+
+        start_step: int = 0
```
Comment (rakkit) [inline torchtitan/components/ema.py:78]: i think this will be less used, mainly to offer the options to avoid code-start case..

File: torchtitan/components/ema.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        begins. Intentionally decoupled from the LR scheduler's WSD phases."""
+
+        step_bias: int = 0
+        """Added to (current_step - start_step) when computing t (only relevant
+        when decay is None). A normal resume already gets correct continuity
+        for free since current_step is restored as usual; step_bias is only
+        for deliberately renumbering Trainer.step (e.g. a new training phase)
+        while wanting the EMA to keep aging as if uninterrupted."""
+
+        update_every_n_steps: int = 1
```
Comment (rakkit) [inline torchtitan/components/ema.py:88]: i think this will be also less used.. i think cost of per-step update is ok.. but we can offer that options anyhow

File: torchtitan/trainer.py [hunk 6]
```diff
@@ -488,6 +493,15 @@ def __init__(self, config: Config):
         self.metrics_processor.optimizers = self.optimizers
         self.metrics_processor.model_parts = self.model_parts
 
+        # Online EMA of model weights (e.g. for cheap mid-WSD-training eval
+        # without a full LR decay). Always built and the hook always
+        # registered; EMAOptimizersContainer internally no-ops everything
+        # when config.ema_weights.enable is False.
+        self.ema_optimizer = config.ema_weights.build(model_parts=self.model_parts)
+        self.optimizers.register_step_post_hook(
```
Comment (rakkit) [inline torchtitan/trainer.py:501]: i think we can also to explicitly call  `     self.ema_optimizer.step()` after optimizers.step,
in later case better add a stream check to make sure ema update always happend after optimizer update done?

File: torchtitan/components/ema.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                for e, p in zip(ema_params, params):
+                    e.copy_(e * decay + p * (1.0 - decay))
+
+    # --- CPU offload path (GH200-optimized: async side-stream, pinned memory) ---
+
+    def _init_cpu_offload(self) -> None:
+        self._offload_stream = torch.cuda.Stream()
+        for ema_opt in self.optimizers:
+            for st in ema_opt.state.values():
+                st["ema_params"] = st["ema_params"].cpu().pin_memory()
```
Comment (NaDongsfd) [inline torchtitan/components/ema.py]: Is EMA CPU offload currently expected to work with FSDP2? _init_cpu_offload() appears to call .pin_memory() directly on a DTensor, which raises NYI: aten._pin_memory.default.

Comment (rakkit) [reply torchtitan/components/ema.py]: from my local test it works good.  (cpu offload is more or less like adamw offload, i add pin memory and stream to make offload works better on GH/GB)

Comment (NaDongsfd) [reply torchtitan/components/ema.py]: Thanks. We used this PR to validate EMA for our work and encountered an issue with FSDP2 CPU offload. After FSDP2 is applied, the EMA states are DTensors. Since .cpu() still returns a DTensor, the subsequent .pin_memory() call fails in our environment with NYI: aten._pin_memory.default.
The PyTorch source at our commit 7661cd9c6b841b62b7f411aa52ec51f05457263b also indicates this limitation: torch/distributed/tensor/parallel/fsdp.py, line 110 states, ignore memory_format and pin_memory as those are not supported by DT.
Is the current implementation expected to support FSDP2 CPU offload when the EMA states are DTensors, or could this behavior depend on the PyTorch build?

Comment (rakkit) [reply torchtitan/components/ema.py]: thanks! ok that's interesting, i will do more test later today and update you later

If DTensor really not gonna work, the minimal fix is to add Dtensor > Tensor convert on saving and dtensor reaload on loading.

Comment (rakkit) [reply torchtitan/components/ema.py]: you are right.   i fixed it by explicitly apply Dtensor -> Tensor converting

File: torchtitan/components/ema.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+        for ema_opt in self.optimizers:
+            for param_state in ema_opt.state.values():
+                param_state["ema_params"] = self._pin_local(param_state["ema_params"])
+
+    def _pin_local(self, tensor: torch.Tensor) -> torch.Tensor:
+        """DTensor has no pin_memory() dispatch support (NYI:
+        aten._pin_memory.default), so pin just the local shard."""
+        return self._local_view(tensor).cpu().pin_memory()
+
+    def _materialize_dtensor(
```
Comment (tianyu-l) [inline torchtitan/components/ema.py:183]: What gap does DTensor have so that we need the back-and-forth conversion here?

File: torchtitan/trainer.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/trainer.py:501]: what can be off-stream / async? I had thought they would happen on the same main stream.

File: torchtitan/components/checkpoint.py [hunk 9]
```diff
@@ -898,12 +905,17 @@ def load(self, step: int = -1) -> bool:
         begin = time.monotonic()
 
         states = self._states_to_load(model_only)
+        needs_ema_reseed = not model_only and self._maybe_exclude_missing_ema(
```
Comment (tianyu-l) [inline torchtitan/components/checkpoint.py:908]: what does `reseed` mean?

**Review by tianyu-l — COMMENTED** [review]
Probably because I don't have a good mental model, I found the logic around two things not clear:
- Checkpointing. Could you describe the concrete requirements around checkpoint save & load, HF conversion expectation, and downstream eval? I only know that people may want ema'ed checkpoints for eval, but there're many other aspects -- e.g. should we allow loading a checkpoint with / without ema?
- Offloading. When exactly would offloading and unloading happen? Would any thing happen in async way?

File: torchtitan/trainer.py (hunk 6 above)
Comment (rakkit) [reply torchtitan/trainer.py:501]: current version has no "async".


> " I had thought they would happen on the same main stream."
yes, by default i think it will be executed after `optmizer.step` done.   
There might be an exception is if we want to have better CPU-offload/fetch -> we need to create stream -> we may want to explicitly wait main stream is done.

i think my original comment may not be related.  doing eithter explicitly ` self.ema_optimizer.step()` or post hook should work

File: torchtitan/components/ema.py (hunk 8 above)
Comment (rakkit) [reply torchtitan/components/ema.py:183]: as @NaDongsfd mentioned above,  DTensor does not have/support `_pin_memory`, so if one need CPU-Offload, we need extra convert there

**Comment by rakkit** [comment]
> Checkpointing. Could you describe the concrete requirements around checkpoint save & load, HF conversion expectation, and downstream eval? I only know that people may want ema'ed checkpoints for eval, but there're many other aspects -- e.g. should we allow loading a checkpoint with / without ema?


For HF conversion & Eval, by default it only convert "main" weights (act like no EMA), and optionally to convert both "main" weights and "EMA" weights when `--ema_output` is given so to get two HF ckpts.   it is for sure optional to export EMA only and without main weights.  
For WSD users its better to do downstream evals on EMA checkpoint (we have tested this PR on 30b model for ~700B tokens and EMA did work well as expected, we could update some eval plots maybe next week)


For checkpoint save/load, as it is like a StatefulOptimizer, the EMA state will be saved just like model weights and optimizer states, and async checkpoint et al works well and out-of-box. (i tested 30b model on up 1024 gpus with HSDP,  async ckpt works well).     For sure, all DCP checkpoints now will be larger.


It will be complicated to have EMA but load a CKPT without EMA weights, so we need to copy mater weights to init the EMA weights, and some corner cases like this 





> Offloading. When exactly would offloading and unloading happen? Would any thing happen in async way?

For now, this PR only do memory offload.  so in training, the EMA copy of weights can be on CPU RAM, and after optimizer step,  EMA weights are load to GPU and do update, and then offload to CPU RAM again.   this is avoid put EMA resident on GPU RAM to save space.   we can off course move EMA update to CPU as well.

File: torchtitan/components/checkpoint.py (hunk 9 above)
Comment (rakkit) [reply torchtitan/components/checkpoint.py:908]: if i remember correctly, there is an edge case in my test.

train model with this PR, with EMA OFF,    the CKPT will have `states[EMA_OPTIMIZER]`  but its empty.

If we resume training with that CKPT, with EMA ON,  `dcp_load` will try to read `states[EMA_OPTIMIZER]` and believe the CKPT do have data there, and it casuses problem because the CKPT comes from runs without EMA.

So if we need to clean the empty dict of `states[EMA_OPTIMIZER]` for `dcp_load`, then, create `states[EMA_OPTIMIZER]` equals to "main" model weights for cold-start

File: torchtitan/components/ema.py (hunk 8 above)
Comment (tianyu-l) [reply torchtitan/components/ema.py:183]: @pianpwk @wconstab are you aware of any difficulty, or is it just not supported?

File: scripts/checkpoint_conversion/convert_to_hf.py [hunk 10]
```diff
@@ -6,14 +6,142 @@
...
+from torch.futures import Future
 from torchtitan.components.checkpoint import ModelWrapper
+from torchtitan.components.checkpoint_utils import canonical_fqn
 from torchtitan.config import TORCH_DTYPE_MAP
 
+_EMA_STATE_PREFIX = "ema_optimizer.state."
+_EMA_STATE_SUFFIX = ".ema_params"
+
+
+class ParallelFileSystemReader(FileSystemReader):
```
Comment (rakkit) [inline scripts/checkpoint_conversion/convert_to_hf.py:29]: this helper is used to speed up DCP -> Convert (i can open separate PR for this if you want).

I dont know how people usually run convert to HF.  i usually run it on single cpu node, therefore when i convert DCP from actual runs (for example 30b on 1k gpus) it takes ages to load the data.  and we can make this faster by just using more cpu cores

File: torchtitan/components/ema.py (hunk 7 above)
Comment (NaDongsfd) [reply torchtitan/components/ema.py]: Sorry, I missed your earlier reply. Thanks for getting back to me. btw, do you have an estimated timeline for merging PR #3985? No worries if there isn’t an update yet.

Comment (rakkit) [reply torchtitan/components/ema.py]: i am not sure on my side, mainly up to PR review.

I have tested it on 30b model on 512 gpus for ~500B tokens and it seems works ok for me (but without cpu-offload)

File: scripts/checkpoint_conversion/convert_to_hf.py (hunk 10 above)
Comment (ruisizhang123) [reply scripts/checkpoint_conversion/convert_to_hf.py:29]: Curious, is there any specific tasks you wanted to evaluate with huggingface format? 

FYI there is this draft pr that allows you to evaluate lm-harness with a torchtitan DCP format without converting to huggingface format: https://github.com/EleutherAI/lm-evaluation-harness/pull/3948

File: torchtitan/components/checkpoint.py [hunk 11]
```diff
@@ -210,8 +212,11 @@ class CheckpointManager(Configurable):
         optimizers (OptimizersContainer): The optimizers used to optimize the model.
         lr_schedulers (LRSchedulersContainer): The lr schedulers used to optimize
             the model.
+        ema_optimizer (EMAOptimizersContainer): Online EMA of model weights.
+            Always built and passed in (see EMAOptimizersContainer), so its
+            checkpoint contribution is empty when EMA is disabled.
         states (Dict[str, Any]): The states that need to be saved, other than the
-            previous 4 components.
+            previous 5 components.
         sd_adapter (Optional[type[BaseStateDictAdapter]]): The adapter used to convert
```
Comment (ruisizhang123) [inline torchtitan/components/checkpoint.py]: Question: From my understanding, we are saving both last model checkpoint and EMA'ed checkpoint in this PR? Shouldn't we only save EMA'ed weight if ema_optimizer is not None?

File: torchtitan/components/checkpoint.py [hunk 12]
```diff
@@ -210,8 +212,11 @@ class CheckpointManager(Configurable):
         optimizers (OptimizersContainer): The optimizers used to optimize the model.
         lr_schedulers (LRSchedulersContainer): The lr schedulers used to optimize
             the model.
+        ema_optimizer (EMAOptimizersContainer): Online EMA of model weights.
+            Always built and passed in (see EMAOptimizersContainer), so its
+            checkpoint contribution is empty when EMA is disabled.
         states (Dict[str, Any]): The states that need to be saved, other than the
-            previous 4 components.
+            previous 5 components.
```
Comment (ruisizhang123) [inline torchtitan/components/checkpoint.py:219]: why it changed to 5 here?

File: torchtitan/components/ema.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+        return 2.0 ** (-1.0 / (self.half_life_fraction * num_updates))
+
+    def _update(self, num_updates: int) -> None:
+        decay = self._decay_at(num_updates)
+        for part_idx, (ema_opt, model) in enumerate(
+            zip(self.optimizers, self.model_parts)
+        ):
+            params = [p for p in model.parameters() if p.requires_grad]
+            if not params:
+                continue
```
Comment (ruisizhang123) [inline torchtitan/components/ema.py:151]: I feel like buffers like expert bias [things](https://github.com/pytorch/torchtitan/blob/ec30db9d095fb2b3e50733a13216c77aa59292b5/torchtitan/models/common/moe.py#L389-L401) should also be EMA'ed. However, they seemed to be missing from this update, since it's only calling parameters()

File: torchtitan/components/checkpoint.py (hunk 9 above)
Comment (ruisizhang123) [reply torchtitan/components/checkpoint.py:908]: hmmm maybe we can make "EMA_OPTIMIZER" optional in state? if there is ema checkpoint, we load it, otherwise; it will be marked as None or sth?

**Review by ruisizhang123 — COMMENTED** [review]
Thanks for adding EMA! have a few questions and comments.

**Comment by tianyu-l** [comment]
@rakkit would you like to follow up or would you like us to take over?

File: torchtitan/components/checkpoint.py (hunk 9 above)
Comment (rakkit) [reply torchtitan/components/checkpoint.py:908]: > hmmm maybe we can make "EMA_OPTIMIZER" optional in state? if there is ema checkpoint, we load it, otherwise; it will be marked as None or sth?

conceptually yes... just need to do lots of patch...

File: scripts/checkpoint_conversion/convert_to_hf.py (hunk 10 above)
Comment (rakkit) [reply scripts/checkpoint_conversion/convert_to_hf.py:29]: thanks a lot,  thats a good idea to use DCP for eval.

i think HF format is not just for "eval", for example for us, we want to build open source model and we wish to release all (intermedia) checkpoints like olmo or other projects.  so make other people can also access the model in easy way.

(and for now our eval run on top of VLLM, its decoupled with torch-titan for now.  it prefer to read HF format of checkpoint)

File: torchtitan/components/checkpoint.py (hunk 12 above)
Comment (rakkit) [reply torchtitan/components/checkpoint.py:219]: before:
```
        self.states.update(
            {
                MODEL: ModelWrapper(model_parts),
                OPTIMIZER: optimizers,
                DATALOADER: dataloader,
                LR_SCHEDULER: lr_schedulers,
            }
        )
        
```

and now one more `EMA_OPTIMIZER: ema_optimizer` state, so its 5.

File: torchtitan/components/ema.py (hunk 13 above)
Comment (rakkit) [reply torchtitan/components/ema.py:151]: good question, i am fine to include it, or make it optional.  

TBH i dont know what is the right way.   because e.g.  people dont do `expert_bias_update_coeff` scheduler or don't decay the coeff.    in this sense we may not need it.  (EMA + constant LR  ≈ LR decay scheduler and if we don't need decay we may not needs RMA)

**Comment by rakkit** [comment]
sry for late reply.... i did not see notification from GitHub so i did not realize you have comments on it

File: scripts/checkpoint_conversion/convert_to_hf.py (hunk 10 above)
Comment (ruisizhang123) [reply scripts/checkpoint_conversion/convert_to_hf.py:29]: i see, thanks for explaining

File: torchtitan/components/ema.py (hunk 13 above)
Comment (ruisizhang123) [reply torchtitan/components/ema.py:151]: Maybe we can have an optional `ema_buffer` key in optimizer, and have some customized regex function if user wants to apply `ema_buffer` to a subset of buffers. wdyt?

Comment (rakkit) [reply torchtitan/components/ema.py:151]: i am fine for this, will add it

File: torchtitan/components/checkpoint.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/components/checkpoint.py:1096]: Can we just use this `exclude_from_loading` flag https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/checkpointer/base.py#L315

File: torchtitan/components/ema.py [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+                self.state[p]["ema_params"] = p.detach().clone()
+
+    def step(self, *args, **kwargs) -> None:
+        raise RuntimeError(
+            "_EMAParamOptimizer must not be step()-ed; call "
+            "EMAOptimizersContainer.step(current_step) instead."
+        )
+
+
+class EMAOptimizersContainer(OptimizersContainer):
```
Comment (tianyu-l) [inline torchtitan/components/ema.py:49]: I'd suggest to make name succinct
```suggestion
class EMA(OptimizersContainer):
```

File: torchtitan/components/ema.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+    ``__init__``/``step()``/``zero_grad()`` -- this is never a real training
+    optimizer. Never merged into ``Trainer.optimizers`` or
+    ``LRSchedulersContainer`` -- it's a sibling object, always built and
+    wired into ``CheckpointManager`` and a ``register_step_post_hook``;
+    ``enable`` decides whether that amounts to anything.
+    """
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(Configurable.Config):
+        enable: bool = False
```
Comment (tianyu-l) [inline torchtitan/components/ema.py:63]: we don't need this field -- `trainer.ema` config being `None` means it's not enabled

File: torchtitan/components/ema.py [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+    def _init_cpu_offload(self) -> None:
+        self._offload_stream = torch.cuda.Stream()
+        for ema_opt in self.optimizers:
+            for param_state in ema_opt.state.values():
+                param_state["ema_params"] = self._pin_local(param_state["ema_params"])
+
+    def _pin_local(self, tensor: torch.Tensor) -> torch.Tensor:
+        """DTensor has no pin_memory() dispatch support (NYI:
+        aten._pin_memory.default), so pin just the local shard."""
+        return self._local_view(tensor).cpu().pin_memory()
```
Comment (tianyu-l) [inline torchtitan/components/ema.py:181]: please add a TODO for DTensor to support `pin_memory` -- there might be some challenge on PG restore
cc @sanketpurandare

File: torchtitan/trainer.py [hunk 17]
```diff
@@ -221,6 +208,7 @@ def maybe_log(self) -> None:
     loss_fn: BaseLoss
     optimizers: OptimizersContainer
     lr_schedulers: LRSchedulersContainer
+    ema_optimizer: EMAOptimizersContainer
```
Comment (tianyu-l) [inline torchtitan/trainer.py:211]:
```suggestion
    ema: EMA
```

File: torchtitan/trainer.py [hunk 18]
```diff
@@ -91,6 +92,9 @@ class Config(Configurable.Config):
         lr_scheduler: LRSchedulersContainer.Config = field(
             default_factory=LRSchedulersContainer.Config
         )
+        ema_weights: EMAOptimizersContainer.Config = field(
```
Comment (tianyu-l) [inline torchtitan/trainer.py:95]:
```suggestion
        ema: EMA.Config = field(
```

File: torchtitan/trainer.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/trainer.py:501]: let's just call `self.ema.step()`

File: torchtitan/components/checkpoint.py
Comment (tianyu-l) [inline torchtitan/components/checkpoint.py:1]: > For checkpoint save/load, as it is like a StatefulOptimizer, the EMA state will be saved just like model weights and optimizer states, and async checkpoint et al works well and out-of-box. (i tested 30b model on up 1024 gpus with HSDP, async ckpt works well). For sure, all DCP checkpoints now will be larger.

Let's support this first.

> It will be complicated to have EMA but load a CKPT without EMA weights, so we need to copy mater weights to init the EMA weights, and some corner cases like this

I think it'd be fine? We just don't skip loading ema weights, and start to accumulate ema values using the current main weights.

> For HF conversion & Eval, by default it only convert "main" weights (act like no EMA), and optionally to convert both "main" weights and "EMA" weights when --ema_output is given so to get two HF ckpts.

For using ema to do checkpoint eval, in this PR I'd suggest
- we don't support "cleanly saving an ema checkpoint to do eval"
- and instead user can have an offline script to convert current "mixed main + ema" DCP checkpoint into an ema-only model checkpoint

And let's worry about that in next PRs.

File: scripts/checkpoint_conversion/convert_to_hf.py (hunk 10 above)
Comment (tianyu-l) [reply scripts/checkpoint_conversion/convert_to_hf.py:29]: Let's separate changes in this file in a separate PR.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

**Comment by tianyu-l** [comment]
@rakkit would you like to follow up?

File: torchtitan/components/checkpoint.py (hunk 9 above)
Comment (rakkit) [reply torchtitan/components/checkpoint.py]: done, its optional now.

File: torchtitan/trainer.py (hunk 6 above)
Comment (rakkit) [reply torchtitan/trainer.py]: fixed

File: torchtitan/components/ema.py (hunk 8 above)
Comment (rakkit) [reply torchtitan/components/ema.py:228]: i add todo about dtensor's to local function

**Comment by rakkit** [comment]
most of comments adressed, thx for feedback.
here is summary:

-  it is "EMA" now, and its optionally now, i.e `EMA.Config | None = None` in Trainer.Config, 
- we can do --checkpoint.exclude_from_loading=ema, for case of re-loading ckpt without ema
- no hook, and its `self.ema.step(self.step)` now
- buffer is supported and configurable.  for exmaple `config.ema = EMA.Config(buffer_patterns=[r"expert_bias_E$"])`
or  `ema:config --ema.buffer-patterns 'expert_bias_E$'`

File: torchtitan/components/checkpointer/dcp.py [hunk 19]
```diff
@@ -602,6 +613,13 @@ def _load(self, step: int = -1) -> bool:
             from_hf=from_hf,
             from_quantized=from_quantized,
         )
+        # EMA is excluded from loading (e.g. resuming with EMA newly enabled
+        # against a checkpoint that predates it, or was saved with EMA off)
+        # via the same exclude_from_loading a caller uses for any other
+        # optional state -- reseed it from the just-loaded model weights
+        # rather than leaving it at its pre-load (fresh-init) values.
+        if not model_only and EMA in self.exclude_from_loading and EMA in self.states:
+            self.states[EMA].load_state_dict({})
```
Comment (tianyu-l) [inline torchtitan/components/checkpointer/dcp.py:622]: oh I see. Even if the checkpoint has no EMA states, and we specify EMA as being excluded from loading, we still have to init EMA states using loaded model states.

File: torchtitan/components/checkpointer/dcp.py [hunk 20]
```diff
@@ -602,6 +613,13 @@ def _load(self, step: int = -1) -> bool:
             from_hf=from_hf,
             from_quantized=from_quantized,
         )
+        # EMA is excluded from loading (e.g. resuming with EMA newly enabled
+        # against a checkpoint that predates it, or was saved with EMA off)
+        # via the same exclude_from_loading a caller uses for any other
+        # optional state -- reseed it from the just-loaded model weights
+        # rather than leaving it at its pre-load (fresh-init) values.
+        if not model_only and EMA in self.exclude_from_loading and EMA in self.states:
```
Comment (tianyu-l) [inline torchtitan/components/checkpointer/dcp.py:621]: Why conditioning on `model_only`? e.g. when loading from HuggingFace format, it will be `model_only`, but it doesn't mean we shouldn't set initial EMA states to be model states.

File: torchtitan/components/ema.py
Comment (tianyu-l) [inline torchtitan/components/ema.py:1]: Given how close they are, can we put this under `components/optimizer/`?

**Comment by tianyu-l** [comment]
@claude review the latest commit and my latest comments

**Comment by rakkit** [comment]
thanks a lot, i will fix. (It so annoying i cannot see github notification for no reason, but thanks a lot for your time and quick review)

**Comment by rakkit** [comment]
fixed first bug claude pointed, and rebased to latest commits. 
ema now moved to `components/optimizer/` as claude suggested.
for PP, i think we handel it correctly 

the checkpoint build now needs to take  `ema: EMAContainer | None,` so i explicitly export `ema=None` in all trainer where we need to build checkpoint

File: tests/unit_tests/test_ema.py
Comment (tianyu-l) [inline tests/unit_tests/test_ema.py:1]: need to put under unit_tests/cpu/

File: torchtitan/components/checkpointer/dcp.py [hunk 21]
```diff
@@ -45,12 +45,17 @@
 
 logger = logging.getLogger(__name__)
 
+# "EMA" is also the class name (torchtitan.components.optimizer.ema.EMA), aliased on
+# import below to avoid shadowing this state-dict key constant, which follows
+# the same NAME = "name" convention as MODEL/OPTIMIZER/etc. above.
+EMA = "ema"
```
Comment (tianyu-l) [inline torchtitan/components/checkpointer/dcp.py]: why don't we put here? https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/checkpointer/base.py#L37

File: torchtitan/experiments/rl/actors/trainer.py
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py:1]: could you rebase? I think RL trainer would benefit from this field, too, if we put EMA in `TrainingEngine`

File: torchtitan/components/checkpointer/dcp.py
Comment (tianyu-l) [inline torchtitan/components/checkpointer/dcp.py:1]: cc @ivy-zhou for torch_checkpointing support

**Comment by rakkit** [comment]
thx. rebased and did few more round of checks. 

after rebase, ema build is log now under `_initialize_optimizer`

```
    def _initialize_optimizer(self, model_spec: ModelSpec) -> None:
        """Construct optimizers, learning-rate schedulers and the weight EMA."""
        self.optimizers = self.config.optimizer.build(model_parts=self.model_parts)
        if model_spec.post_optimizer_build_fn is not None:
            model_spec.post_optimizer_build_fn(
                self.optimizers, self.model_parts, self.parallel_dims
            )
        self.lr_schedulers = self.config.lr_scheduler.build(
            optimizers=self.optimizers,
            training_steps=self.config.training.steps,
        )
        self.ema = (
            self.config.ema.build(model_parts=self.model_parts)
            if self.config.ema is not None
            else None
        )
```


and, again, as `CheckpointManager` now expect `ema: EMAContainer | None,`, no defualt value, we need to populate ema build everywhere.


I also rewrote the PR message, includes claude summarized few edge-cases

File: torchtitan/experiments/rl/actors/trainer.py
Comment (rakkit) [reply torchtitan/experiments/rl/actors/trainer.py:1]: done

**Comment by rakkit** [comment]
i also bring EMA weights -> HF convert scripts back, the args is mirror of DCP->HF.  its much clean than put eveything into one.


Also i just realized,  torchtitan by default have `last_save_model_only=True`, that means the last checkpoint will have model weights only.  for users of EMA, i think they can safely resolve this by set `last_save_model_only=False`. i am not sure if we wanna to condition on ema for `last_save_model_only=True` behaviors.

**Review by tianyu-l — APPROVED** [review]

**Comment by tianyu-l** [comment]
@rakkit Seems CI failures are real. Also sorry need to rebase again.

**Comment by rakkit** [comment]
rebased to `b9e1d1f` and fixed lint error.
